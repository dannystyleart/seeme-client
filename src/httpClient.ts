import { URLSearchParams } from 'url';
import fetch from 'node-fetch';
import { removeSlash } from './utils';

export const createApiClient = (apiUrl: string) => {
  return <T>(path: string, params: any): Promise<T> => {
    const apiEndpoint = [removeSlash(apiUrl), removeSlash(path, false)].join('/');
    const queryParams = new URLSearchParams(<any>params);

    return new Promise(async (resolve, reject) => {
      try {
        const response = await fetch([apiEndpoint, queryParams.toString()].join('?'));
        resolve(response.json());
      } catch (e) {
        reject(e);
      }
    });
  };
};
