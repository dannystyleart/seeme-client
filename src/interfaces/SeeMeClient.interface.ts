export interface SeeMeClient {
  getBalance(): Promise<any>;
  sendMessage(address: string, message: string): Promise<any>;
}
