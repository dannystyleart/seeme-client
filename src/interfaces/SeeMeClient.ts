import { SendMessageCallbackStatus } from '../SeeMeClient';

export interface SeeMeClient {
  getBalance(): Promise<any>;
  setIP(ipAddress: string): Promise<any>;
  sendSMS(number: string, message: string, optionalParams?: SendMessageOptionalParams): Promise<any>;
}

export interface SendMessageOptionalParams {
  sender?: string;
  referenceId?: string | number;
  callbackStatus?: 'all' | SendMessageCallbackStatus | Array<SendMessageCallbackStatus>;
  callbackURL?: string;
}
