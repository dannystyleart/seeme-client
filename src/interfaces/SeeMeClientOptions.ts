export interface SeeMeClientOptions {
  apiKey: string;
  apiHost?: string;
  apiPath?: string;
  apiVersion?: string;
}
