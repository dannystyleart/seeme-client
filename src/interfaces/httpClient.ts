export interface HttpClientRequest {
  (path: string, params?: any): Promise<any>;
}
