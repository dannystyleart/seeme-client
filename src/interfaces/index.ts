export * from './SeeMeClient';
export * from './SeeMeClientOptions';
