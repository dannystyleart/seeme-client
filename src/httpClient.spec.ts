import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { HttpClientRequest } from './interfaces/httpClient';
import { createApiClient } from './httpClient';

const mockApiUrl = 'http://api.example.com';
const server = setupServer(
  rest.get(`${mockApiUrl}/*`, (req, res, context) => {
    return res(
      context.json({
        message: 'Reponse from mock server',
        requestedUrl: req.url.toString()
      })
    );
  })
);

describe('httpClient', () => {
  beforeAll(() => {
    server.listen();
  });
  afterAll(() => {
    server.close();
  });

  test('createApiClient(apiUrl) should return a request method', () => {
    expect(createApiClient(mockApiUrl)).toBeInstanceOf(Function);
  });

  describe('request(path, params)', () => {
    let request: HttpClientRequest;
    beforeEach(() => {
      request = createApiClient(mockApiUrl);
    });

    test('should return parsed json response sent to given api endpoint defined by path argument', async () => {
      const result = await (<any>request('/hello'));
      expect(result).toMatchSnapshot();
    });

    test('should return parsed response sent to given api endpoint defined by path and params arguments', async () => {
      const result = await (<any>request('/hello', {
        search: 'some search term',
        flagged: true,
        someList: [1, 2, 3]
      }));
      expect(result).toMatchSnapshot();
    });

    test('should not append query params when params is empty object', async () => {
      const result = await (<any>request('/hello', {}));
      expect(result).toMatchSnapshot();
    });

    test('should fail the promise when server response with error', async () => {
      server.use(
        rest.get(`${mockApiUrl}/failing-api`, (req, res) => {
          return res.networkError('Something failed');
        })
      );

      const failHandler = jest.fn();
      await request('/failing-api').catch(failHandler);
      expect(failHandler).toHaveBeenCalled();
    });
  });
});
