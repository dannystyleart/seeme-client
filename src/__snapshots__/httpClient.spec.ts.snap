// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`httpClient request(path, params) should not append query params when params is empty object 1`] = `
Object {
  "message": "Reponse from mock server",
  "requestedUrl": "http://api.example.com/hello",
}
`;

exports[`httpClient request(path, params) should return parsed json response sent to given api endpoint defined by path argument 1`] = `
Object {
  "message": "Reponse from mock server",
  "requestedUrl": "http://api.example.com/hello",
}
`;

exports[`httpClient request(path, params) should return parsed response sent to given api endpoint defined by path and params arguments 1`] = `
Object {
  "message": "Reponse from mock server",
  "requestedUrl": "http://api.example.com/hello?search=some+search+term&flagged=true&someList=1%2C2%2C3",
}
`;
